File: src/main/java/org/radarbase/fcm/downstream/AdminSdkFcmSender.java
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */

package org.radarbase.fcm.downstream;

import com.google.auth.oauth2.GoogleCredentials;
import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import com.google.firebase.messaging.AndroidConfig;
import com.google.firebase.messaging.AndroidConfig.Priority;
import com.google.firebase.messaging.AndroidNotification;
import com.google.firebase.messaging.ApnsConfig;
import com.google.firebase.messaging.Aps;
import com.google.firebase.messaging.ApsAlert;
import com.google.firebase.messaging.FcmOptions;
import com.google.firebase.messaging.FirebaseMessaging;
import com.google.firebase.messaging.FirebaseMessagingException;
import com.google.firebase.messaging.Message;
import com.google.firebase.messaging.Notification;
import java.io.IOException;
import java.time.Duration;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import lombok.extern.slf4j.Slf4j;
import org.radarbase.fcm.model.FcmDataMessage;
import org.radarbase.fcm.model.FcmDownstreamMessage;
import org.radarbase.fcm.model.FcmNotificationMessage;

/**
 * When authorizing via a service account, you have to set the GOOGLE_APPLICATION_CREDENTIALS
 * environment variable. For More info, see
 * https://firebase.google.com/docs/admin/setup#initialize-sdk
 *
 * @author yatharthranjan
 */
@Slf4j
@SuppressWarnings({"PMD.DataflowAnomalyAnalysis", "PMD.AvoidDuplicateLiterals"})
public class AdminSdkFcmSender implements FcmSender {
  static final int DEFAULT_TIME_TO_LIVE = 2_419_200; // 28 days in seconds

  public AdminSdkFcmSender() throws IOException {
    // TODO also take config from application properties
    FirebaseOptions options =
        new FirebaseOptions.Builder()
            .setCredentials(GoogleCredentials.getApplicationDefault())
            .build();

    try {
      FirebaseApp.initializeApp(options);
    } catch (IllegalStateException exc) {
      log.warn("Firebase app was already initialised. {}", exc.getMessage());
    }
  }

  @Override
  public void send(FcmDownstreamMessage downstreamMessage) throws FirebaseMessagingException {
    // TODO Add support for WebPush as well
    String priority = downstreamMessage.getPriority();

    Message.Builder message =
        Message.builder()
            .setToken(downstreamMessage.getTo())
            .setFcmOptions(FcmOptions.builder().build())
            .setCondition(downstreamMessage.getCondition());

    int ttl = getValidTtlMillis(downstreamMessage.getTimeToLive());

    if (downstreamMessage instanceof FcmNotificationMessage) {
      FcmNotificationMessage notificationMessage = (FcmNotificationMessage) downstreamMessage;

      message
          .setAndroidConfig(
              AndroidConfig.builder()
                  .setCollapseKey(downstreamMessage.getCollapseKey())
                  .setPriority(priority == null ? Priority.HIGH : Priority.valueOf(priority))
                  .setTtl(ttl)
                  .setNotification(getAndroidNotification(notificationMessage))
                  .putAllData(notificationMessage.getData())
                  .build())
          .setApnsConfig(
              getApnsConfigBuilder(notificationMessage, ttl)
                  .putHeader("apns-push-type", "alert")
                  .build())
          .putAllData(notificationMessage.getData())
          .setCondition(notificationMessage.getCondition())
          .setNotification(
              Notification.builder()
                  .setBody(
                      String.valueOf(
                          notificationMessage.getNotification().getOrDefault("body", "")))
                  .setTitle(
                      String.valueOf(
                          notificationMessage.getNotification().getOrDefault("title", "")))
                  .setImage(
                      String.valueOf(
                          notificationMessage.getNotification().getOrDefault("image_url", "")))
                  .build());
    } else if (downstreamMessage instanceof FcmDataMessage) {
      FcmDataMessage dataMessage = (FcmDataMessage) downstreamMessage;
      message
          .setAndroidConfig(
              AndroidConfig.builder()
                  .setCollapseKey(downstreamMessage.getCollapseKey())
                  .setPriority(priority == null ? Priority.NORMAL : Priority.valueOf(priority))
                  .setTtl(ttl)
                  .putAllData(dataMessage.getData())
                  .build())
          .setApnsConfig(getApnsConfigBuilder(dataMessage, ttl).build())
          .setCondition(dataMessage.getCondition())
          .putAllData(dataMessage.getData());
    } else {
      throw new IllegalArgumentException(
          "The Message type is not known." + downstreamMessage.getClass());
    }

    String response = FirebaseMessaging.getInstance().send(message.build());
    log.info("Message Sent with response : {}", response);
  }

  private AndroidNotification getAndroidNotification(FcmNotificationMessage notificationMessage) {
    AndroidNotification.Builder builder =
        AndroidNotification.builder()
            .setBody(String.valueOf(notificationMessage.getNotification().getOrDefault("body", "")))
            .setTitle(
                String.valueOf(notificationMessage.getNotification().getOrDefault("title", "")))
            .setChannelId(
                getString(notificationMessage.getNotification().get("android_channel_id")))
            .setColor(getString(notificationMessage.getNotification().get("color")))
            .setTag(getString(notificationMessage.getNotification().get("tag")))
            .setIcon(getString(notificationMessage.getNotification().get("icon")))
            .setSound(getString(notificationMessage.getNotification().get("sound")))
            .setClickAction(getString(notificationMessage.getNotification().get("click_action")));

    String bodyLocKey = getString(notificationMessage.getNotification().get("body_loc_key"));
    String titleLocKey = getString(notificationMessage.getNotification().get("title_loc_key"));

    if (bodyLocKey != null) {
      builder
          .setBodyLocalizationKey(
              getString(notificationMessage.getNotification().get("body_loc_key")))
          .addBodyLocalizationArg(
              getString(notificationMessage.getNotification().get("body_loc_args")));
    }

    if (titleLocKey != null) {
      builder
          .addTitleLocalizationArg(
              getString(notificationMessage.getNotification().get("title_loc_args")))
          .setTitleLocalizationKey(
              getString(notificationMessage.getNotification().get("title_loc_key")));
    }

    return builder.build();
  }

  /**
   * Get the APNS config builder. More info on values and keys -
   * https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns/
   * and
   * https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification
   */
  private ApnsConfig.Builder getApnsConfigBuilder(FcmDownstreamMessage message, int ttl) {
    ApnsConfig.Builder config = ApnsConfig.builder();

    if (message.getCollapseKey() != null)
      config.putHeader("apns-collapse-id", message.getCollapseKey());

    // The date at which the notification is no longer valid. This value is a UNIX epoch
    // expressed in seconds (UTC).
    config.putHeader(
        "apns-expiration",
        String.valueOf(Instant.now().plus(Duration.ofMillis(ttl)).toEpochMilli() / 1000));

    if (message instanceof FcmNotificationMessage) {
      FcmNotificationMessage notificationMessage = (FcmNotificationMessage) message;
      Map<String, Object> apnsData = new HashMap<>(notificationMessage.getData());

      ApsAlert.Builder apsAlertBuilder = ApsAlert.builder();
      String title = getString(notificationMessage.getNotification().get("title"));
      if (title != null) apsAlertBuilder.setTitle(title);

      String body = getString(notificationMessage.getNotification().get("body"));
      if (body != null) apsAlertBuilder.setBody(body);

      String titleLocKey = getString(notificationMessage.getNotification().get("title_loc_key"));
      if (titleLocKey != null) apsAlertBuilder.setTitleLocalizationKey(titleLocKey);

      String titleLocArgs = getString(notificationMessage.getNotification().get("title_loc_args"));
      if (titleLocKey != null && titleLocArgs != null)
        apsAlertBuilder.addTitleLocalizationArg(titleLocArgs);

      String bodyLocKey = getString(notificationMessage.getNotification().get("body_loc_key"));
      if (bodyLocKey != null) apsAlertBuilder.setLocalizationKey("body_loc_key");

      String bodyLocArgs = getString(notificationMessage.getNotification().get("body_loc_args"));
      if (bodyLocKey != null && bodyLocArgs != null)
        apsAlertBuilder.addLocalizationArg(bodyLocArgs);

      Aps.Builder apsBuilder = Aps.builder();
      String sound = getString(notificationMessage.getNotification().get("sound"));
      if (sound != null) apsBuilder.setSound(sound);

      String badge = getString(notificationMessage.getNotification().get("badge"));
      if (badge != null) apsBuilder.setBadge(Integer.parseInt(badge));

      String category = getString(notificationMessage.getNotification().get("category"));
      if (category != null) apsBuilder.setCategory(category);

      String threadId = getString(notificationMessage.getNotification().get("thread_id"));
      if (threadId != null) apsBuilder.setThreadId(threadId);

      if (notificationMessage.getContentAvailable() != null)
        apsBuilder.setContentAvailable(notificationMessage.getContentAvailable());

      if (notificationMessage.getMutableContent() != null)
        apsBuilder.setMutableContent(notificationMessage.getMutableContent());

      return config
          .putAllCustomData(apnsData)
          .setAps(apsBuilder.setAlert(apsAlertBuilder.build()).build())
          .putHeader("apns-push-type", "alert");

    } else if (message instanceof FcmDataMessage) {
      FcmDataMessage dataMessage = (FcmDataMessage) message;
      Map<String, Object> apnsData = new HashMap<>(dataMessage.getData());

      return config
          .putAllCustomData(apnsData)
          .setAps(Aps.builder().setContentAvailable(true).setSound("default").build())
          .putHeader("apns-push-type", "background") // No alert is shown
          .putHeader("apns-priority", "5"); // 5 required in case of background type
    } else {
      throw new IllegalArgumentException("The Message type is not known." + message.getClass());
    }
  }

  private String getString(Object obj) {
    return obj == null ? null : String.valueOf(obj);
  }

  @Override
  public boolean doesProvideDeliveryReceipt() {
    return false;
  }

  public int getValidTtlMillis(int ttl) {
    // Converts ttl from seconds to milliseconds and makes sure this is less than 28 days
    int ttlSeconds = ttl >= 0 && ttl <= DEFAULT_TIME_TO_LIVE ? ttl : DEFAULT_TIME_TO_LIVE;
    return ttlSeconds * 1000;
  }

}

File: appserver-legacy/src/main/kotlin/org/radarbase/fcm/downstream/AdminSdkFcmSender.kt
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */
package org.radarbase.fcm.downstream

import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.messaging.AndroidConfig
import com.google.firebase.messaging.AndroidNotification
import com.google.firebase.messaging.ApnsConfig
import com.google.firebase.messaging.Aps
import com.google.firebase.messaging.ApsAlert
import com.google.firebase.messaging.FcmOptions
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.messaging.FirebaseMessagingException
import com.google.firebase.messaging.Message
import com.google.firebase.messaging.Notification
import org.radarbase.fcm.model.FcmDataMessage
import org.radarbase.fcm.model.FcmDownstreamMessage
import org.radarbase.fcm.model.FcmNotificationMessage
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.time.Duration
import java.time.Instant

/**
 * When authorizing via a service account, you have to set the GOOGLE_APPLICATION_CREDENTIALS
 * environment variable. For More info, see
 * https://firebase.google.com/docs/admin/setup#initialize-sdk
 *
 * @author yatharthranjan
 */
class AdminSdkFcmSender(options: FirebaseOptions) : FcmSender {
    init {
        // TODO also take config from application properties
        try {
            FirebaseApp.initializeApp(options)
        } catch (exc: IllegalStateException) {
            log.warn("Firebase app was already initialised. {}", exc.message)
        }
    }

    @Throws(FirebaseMessagingException::class)
    override fun send(downstreamMessage: FcmDownstreamMessage) {
        // TODO Add support for WebPush as well
        val priority = downstreamMessage.priority

        val message =
            Message.builder()
                .setToken(downstreamMessage.to)
                .setFcmOptions(FcmOptions.builder().build())
                .setCondition(downstreamMessage.condition)

        val ttl = getValidTtlMillis(downstreamMessage.timeToLive!!)

        if (downstreamMessage is FcmNotificationMessage) {
            val notificationMessage = downstreamMessage

            message
                .setAndroidConfig(
                    AndroidConfig.builder()
                        .setCollapseKey(downstreamMessage.collapseKey)
                        .setPriority(
                            if (priority == null) {
                                AndroidConfig.Priority.HIGH
                            } else {
                                AndroidConfig.Priority.valueOf(
                                    priority,
                                )
                            },
                        )
                        .setTtl(ttl.toMillis())
                        .setNotification(getAndroidNotification(notificationMessage))
                        .putAllData(notificationMessage.data)
                        .build(),
                )
                .setApnsConfig(
                    getApnsConfigBuilder(notificationMessage, ttl)!!
                        .putHeader("apns-push-type", "alert")
                        .build(),
                )
                .putAllData(notificationMessage.data)
                .setCondition(notificationMessage.condition)
                .setNotification(
                    Notification.builder()
                        .setBody(
                            notificationMessage.notification!!.getOrDefault("body", "").toString(),
                        )
                        .setTitle(
                            notificationMessage.notification!!.getOrDefault("title", "").toString(),
                        )
                        .setImage(
                            notificationMessage.notification!!.getOrDefault("image_url", "").toString(),
                        )
                        .build(),
                )
        } else if (downstreamMessage is FcmDataMessage) {
            val dataMessage = downstreamMessage
            message
                .setAndroidConfig(
                    AndroidConfig.builder()
                        .setCollapseKey(downstreamMessage.collapseKey)
                        .setPriority(
                            if (priority == null) {
                                AndroidConfig.Priority.NORMAL
                            } else {
                                AndroidConfig.Priority.valueOf(
                                    priority,
                                )
                            },
                        )
                        .setTtl(ttl.toMillis())
                        .putAllData(dataMessage.data)
                        .build(),
                )
                .setApnsConfig(getApnsConfigBuilder(dataMessage, ttl)!!.build())
                .setCondition(dataMessage.condition)
                .putAllData(dataMessage.data)
        } else {
            throw IllegalArgumentException(
                "The Message type is not known." + downstreamMessage.javaClass,
            )
        }

        val response = FirebaseMessaging.getInstance().send(message.build())
        log.info("Message Sent with response : {}", response)
    }

    private fun getAndroidNotification(notificationMessage: FcmNotificationMessage): AndroidNotification? {
        val builder =
            AndroidNotification.builder()
                .setBody(notificationMessage.notification!!.getOrDefault("body", "").toString())
                .setTitle(
                    notificationMessage.notification!!.getOrDefault("title", "").toString(),
                )
                .setChannelId(
                    getString(notificationMessage.notification!!["android_channel_id"]),
                )
                .setColor(getString(notificationMessage.notification!!["color"]))
                .setTag(getString(notificationMessage.notification!!["tag"]))
                .setIcon(getString(notificationMessage.notification!!["icon"]))
                .setSound(getString(notificationMessage.notification!!["sound"]))
                .setClickAction(getString(notificationMessage.notification!!["click_action"]))

        val bodyLocKey = getString(notificationMessage.notification!!["body_loc_key"])
        val titleLocKey = getString(notificationMessage.notification!!["title_loc_key"])

        if (bodyLocKey != null) {
            builder
                .setBodyLocalizationKey(
                    getString(notificationMessage.notification!!["body_loc_key"]),
                )
                .addBodyLocalizationArg(
                    getString(notificationMessage.notification!!["body_loc_args"]),
                )
        }

        if (titleLocKey != null) {
            builder
                .addTitleLocalizationArg(
                    getString(notificationMessage.notification!!["title_loc_args"]),
                )
                .setTitleLocalizationKey(
                    getString(notificationMessage.notification!!["title_loc_key"]),
                )
        }

        return builder.build()
    }

    /**
     * Get the APNS config builder. More info on values and keys -
     * https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns/
     * and
     * https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification
     */
    private fun getApnsConfigBuilder(message: FcmDownstreamMessage, ttl: Duration): ApnsConfig.Builder? {
        val config = ApnsConfig.builder()

        if (message.collapseKey != null) config.putHeader("apns-collapse-id", message.collapseKey)

        // The date at which the notification is no longer valid. This value is a UNIX epoch
        // expressed in seconds (UTC).
        config.putHeader(
            "apns-expiration",
            Instant.now().plus(ttl).epochSecond.toString(),
        )

        if (message is FcmNotificationMessage) {
            val notificationMessage = message
            val apnsData: Map<String?, Any?> = HashMap(notificationMessage.data ?: emptyMap<String, String>())

            val apsAlertBuilder = ApsAlert.builder()
            val title = getString(notificationMessage.notification!!["title"])
            if (title != null) apsAlertBuilder.setTitle(title)

            val body = getString(notificationMessage.notification!!["body"])
            if (body != null) apsAlertBuilder.setBody(body)

            val titleLocKey = getString(notificationMessage.notification!!["title_loc_key"])
            if (titleLocKey != null) apsAlertBuilder.setTitleLocalizationKey(titleLocKey)

            val titleLocArgs = getString(notificationMessage.notification!!["title_loc_args"])
            if (titleLocKey != null && titleLocArgs != null) apsAlertBuilder.addTitleLocalizationArg(titleLocArgs)

            val bodyLocKey = getString(notificationMessage.notification!!["body_loc_key"])
            if (bodyLocKey != null) apsAlertBuilder.setLocalizationKey(bodyLocKey)

            val bodyLocArgs = getString(notificationMessage.notification!!["body_loc_args"])
            if (bodyLocKey != null && bodyLocArgs != null) apsAlertBuilder.addLocalizationArg(bodyLocArgs)

            val apsBuilder = Aps.builder()
            val sound = getString(notificationMessage.notification!!["sound"])
            if (sound != null) apsBuilder.setSound(sound)

            val badge = getString(notificationMessage.notification!!["badge"])
            if (badge != null) apsBuilder.setBadge(badge.toInt())

            val category = getString(notificationMessage.notification!!["category"])
            if (category != null) apsBuilder.setCategory(category)

            val threadId = getString(notificationMessage.notification!!["thread_id"])
            if (threadId != null) apsBuilder.setThreadId(threadId)

            if (notificationMessage.contentAvailable != null) apsBuilder.setContentAvailable(notificationMessage.contentAvailable!!)

            if (notificationMessage.mutableContent != null) apsBuilder.setMutableContent(notificationMessage.mutableContent!!)

            return config
                .putAllCustomData(apnsData)
                .setAps(apsBuilder.setAlert(apsAlertBuilder.build()).build())
                .putHeader("apns-push-type", "alert")
        } else if (message is FcmDataMessage) {
            val dataMessage = message
            val apnsData: Map<String?, Any?> = HashMap(dataMessage.data ?: emptyMap<String, String>())

            return config
                .putAllCustomData(apnsData)
                .setAps(Aps.builder().setContentAvailable(true).setSound("default").build())
                .putHeader("apns-push-type", "background") // No alert is shown
                .putHeader("apns-priority", "5") // 5 required in case of background type
        } else {
            throw IllegalArgumentException("The Message type is not known." + message.javaClass)
        }
    }

    private fun getString(obj: Any?): String? {
        return obj?.toString()
    }

    override fun doesProvideDeliveryReceipt(): Boolean {
        return false
    }

    fun getValidTtlMillis(ttl: Int): Duration {
        // Makes sure ttl is less than 28 days
        val ttlSeconds = if (ttl >= 0 && ttl <= DEFAULT_TIME_TO_LIVE) ttl else DEFAULT_TIME_TO_LIVE
        return Duration.ofSeconds(ttlSeconds.toLong())
    }

    companion object {
        const val DEFAULT_TIME_TO_LIVE: Int = 2419200 // 4 weeks
        private val log: Logger = LoggerFactory.getLogger(AdminSdkFcmSender::class.java)
    }
}
